swupd/rename.go

```diff
@@ -0,0 +1,183 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package swupd
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+)
+
+// BSDIFFSIZE is smallest size of file to consider using bsdiff
+const BSDIFFSIZE = 200
+
+func renameDetection(manifest *Manifest, added []*File, removed []*File, c config) {
+	if len(added) == 0 || len(removed) == 0 {
+		return // nothing to rename
+	}
+	added = trimrenamed(added) // Make copies of input slices, tidy up whilst we are here
+	removed = trimrenamed(removed)
+	if err := fixupstatfields(removed, manifest, &c); err != nil {
+		panic(err)
+	}
+	// Handle pure renames first, don't need to worry about size. Should we skip zero size?
+	// just add call to trimsmall if so
+	// Sort by Hash. No need to have tiebreaker on name as the Manifest links by hash
+	// so 2 identical files being renamed, e.g. python2.6/foo and python3.7/bar being
+	// renamed to python2.7/foo and python3.7/bar have no concept which is the source for the
+	// python2.7/foo. This decision might need to change if the way renames are shown in
+	// the manifest changes.
+	sort.Slice(added, func(i, j int) bool {
+		return added[i].Hash < added[j].Hash
+	})
+	sort.Slice(removed, func(i, j int) bool {
+		return removed[i].Hash < removed[j].Hash
+	})
+	for ax, rx := 0, 0; ax < len(added) && rx < len(removed); {
+		af := added[ax]
+		rf := removed[rx]
+		switch {
+		case af.Hash < rf.Hash:
+			ax++
+		case af.Hash > rf.Hash:
+			rx++
+		default: // Equal hash, so link
+			linkrenamepair(af, rf)
+			ax++
+			rx++
+		}
+	}
+	// Link things with the same names skippingdigits
+	// First remove small files (not worth sending diff) and files which have
+	// exact match
+	added = trimsmall(trimrenamed(added), BSDIFFSIZE)     // Make it explicit we are doing two steps
+	removed = trimsmall(trimrenamed(removed), BSDIFFSIZE) // Can trivially make it one pass.
+	//generate the pairs of *File and short name
+	pa := makepairs(added)
+	pr := makepairs(removed)
+	// Merge where short names match
+	for ax, rx := 0, 0; ax < len(pa) && rx < len(pr); {
+		af := pa[ax]
+		rf := pr[rx]
+		switch {
+		case af.n < rf.n:
+			ax++
+		case af.n > rf.n:
+			rx++
+		default: // Equal truncated name
+			linkrenamepair(af.f, rf.f)
+			ax++
+			rx++
+		}
+	}
+}
+
+// linkrenamepair links two files together
+func linkrenamepair(a, b *File) {
+	a.DeltaPeer = b
+	b.DeltaPeer = a
+	a.Rename = true
+	b.Rename = true
+}
+
+// trimrenamed returns an slice which has had files with DeltaPeers purged
+func trimrenamed(a []*File) []*File {
+	r := make([]*File, 0, len(a))
+	for _, f := range a {
+		if f.Status == StatusGhosted {
```

Comment:
Interesting. Perhaps I need to think a little more about what a ghosted file can be.

We certainly need to add a test case for this.

My reasoning is that a ghosted file shouldn't be a candidate for a renamefrom, as the lifetime of the file is not managed by swupd-client. In particular a ghosted file could be deleted a minute after it is installed. I agree that a `swupd verify --fix` would find it missing, and reinstall it, which could be annoying. An example I can think of is kernel modules being removed from /usr/lib/modules when they have been added to an initrd, which is stored on another partition.

Since you added the ghosted code, do you think that there is a requirement for the ghosted files to exist until there is an update? Put it another way, do you think 'ghosted' means that the files are already dead, or just that we don't have to kill them?

Of course we shouldn't panic here, just skip the file.

Reply:
The whole point of the ghosted file is to enable renames for kernel files. I think currently we actually deprecate old ghosted files immediately after every build, which is the wrong answer (I will add a unit test to check this and fix if necessary). In that case the client would attempt a retry but fail and proceed as normal - we do know that is the current state of the client.

Ghosted only means we don't have to kill them on the client.